app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,174 @@
+    fun get(key: K): T? = synchronized(this) {
+        byKey[key]
+    }
+
+    /** @return the items with the given [keys] that are in the cache */
+    fun getAll(keys: Collection<K>): List<T> = synchronized(this) {
+        keys.mapNotNull { byKey[it] }
+    }
+
+    /**
+     * Removes the keys in [deleted] from cache
+     * Puts the [updatedOrAdded] items into cache only if the containing tile is already cached
+     */
+    fun update(updatedOrAdded: Iterable<T> = emptyList(), deleted: Iterable<K> = emptyList()) = synchronized(this) {
+        for (key in deleted) {
+            val item = byKey.remove(key) ?: continue
+            byTile[item.getTilePos()]?.remove(item)
+        }
+        // items must be removed before they are re-added because e.g. position may have changed
+        for (item in updatedOrAdded) {
+            val oldItem = byKey.remove(item.getKey()) ?: continue
+            byTile[oldItem.getTilePos()]?.remove(oldItem)
+        }
+        // items are only added if the tile they would be sorted into is already cached, because all
+        // tiles that are in cache must be complete
+        for (item in updatedOrAdded) {
+            val tile = byTile[item.getTilePos()]
+            if (tile != null) {
+                tile.add(item)
+                byKey[item.getKey()] = item
+            }
+        }
+    }
+
+    /**
+     * Replaces all tiles fully contained in the [bbox] with the given [items].
+     * Any tile only partially contained in the [bbox] is removed (because all tiles cached must be
+     * complete).
+     * If the number of tiles exceeds maxTiles, only the first maxSize tiles are cached.
+     */
+    fun replaceAllInBBox(items: Collection<T>, bbox: BoundingBox) = synchronized(this) {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val (completelyContainedTiles, incompleteTiles) = tiles.partition { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
+        if (incompleteTiles.isNotEmpty()) {
+            Log.w(TAG, "bbox does not align with tiles, clearing incomplete tiles from cache")
+            for (tile in incompleteTiles) {
+                removeTile(tile)
+            }
+        }
+        replaceAllInTiles(items, completelyContainedTiles)
+
+        trim()
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(items: Collection<T>, tiles: Collection<TilePos>) {
+        // create / replace tiles
+        for (tile in tiles) {
+            removeTile(tile)
+            byTile[tile] = HashSet()
+        }
+
+        val bboxByTile = byTile.keys.associateWith { it.asBoundingBox(tileZoom) }
+        items@for (item in items) {
+            val pos = item.getPosition()
+            for (tile in tiles) {
+                if (pos in bboxByTile[tile]!!) {
+                    byTile[tile]!!.add(item)
+                    byKey[item.getKey()] = item
+                    continue@items
+                }
+            }
```

Comment:
This has changed a bit over time...
* Initially I had used the cited version, with the difference that I only looped over "new" tiles, which I found to be faster than using `getTilePos()`.
* Later I noticed that actually it's necessary to loop over _all_ tiles, because otherwise the download padding causes issues and empty tiles surrounding the downloaded area would wrongly stay empty.
But looping over all tiles provided no advantage over using `getTilePos()`, and may be slower if cache size (tiles) increases. This is now even more relevant as the number of tiles isn't limited any more (theoretically there could be thousands of empty tiles)
* So I had tried going back to `getTilePos()` but that didn't work out because tests started to fail, see https://github.com/streetcomplete/StreetComplete/pull/4125#discussion_r970402242

Reply:
> (I noticed that the commented-out code you removed at my request did not have the check that the item is in one of the tiles passed as a parameter, so the commented-out code was actually erroneous).

Hmm, yes and no. As pointed out above, it putting elements to other tiles turned out to be necessary.
And since I didn't change the name of the function after realizing this, the function now has a very misleading name...

Reply:
Wait, you put data into "incomplete" tiles?

Reply:
It's not incomplete as in the bbox doesn't align with tile edges.

It's incomplete only if you compare it to downloading the same tile.
But it's exactly what is in database after the download, and exactly how the "incomplete" tile is fetched from database.
See commit: https://github.com/streetcomplete/StreetComplete/pull/4125/commits/86731f1c977fde20136704665c99df97d3c0c5e5

Reply:
Hmm, but is see that `getTilePos` is actually used in `update`...

Reply:
So since `getTilePos()` may return the wrong tile, we may need to loop over tiles in `update`, which would probably be really bad for performance...

Reply:
Ok, currently I think using `getTilePos` is ok after all.
If there is a bug causing the test to fail it's most likely in `BoundingBox.enclosingTilesRect`:
`LatLon(0.0001, 0.0001).enclosingBoundingBox(0.0).enclosingTilesRect(16)` just works
`LatLon(0.0, 0.0).enclosingBoundingBox(0.0).enclosingTilesRect(16)` crashes with "IllegalArgumentException: Failed requirement" in TilesRect.kt:52

Reply:
Sorry if I cant follow completely. Regarding **current** implementation:

1. So what exactly do you mean with "incomplete" tiles?  
The download bbox is larger than the the tiles rect that would correspond to it because it has that padding. Now, the elements in that padding, are they put into the cache too?

2. What do you mean with `getTilePos()` may return the wrong tile?

Reply:
> What do you mean with getTilePos() may return the wrong tile?

see https://github.com/streetcomplete/StreetComplete/pull/4125#discussion_r970402242 and below:
> ```kotlin
> val tilesRect = listOf(LatLon(0.0, 0.0), LatLon(0.0002, 0.0002)).enclosingBoundingBox().enclosingTilesRect(16)
> val tilePos = LatLon(0.0, 0.0).enclosingTilePos(16)
> assertTrue(tilesRect.contains(tilePos.toTilesRect()))
> ```

fails, which means that `enclosingTilesRect` and `enclosingTilePos` are inconsistent. This happens when the `LatLon` is exactly at the tile edge, e.g. in case of `LatLon(0.0, 0.0)`.

> So what exactly do you mean with "incomplete" tiles?

I'm not sure why you ask me, as you brought the term "incomplete" into this discussion.
In the code, I use "incomplete tiles" for tiles that do not align with bbox edges, checked with
```kotlin
val tiles = bbox.asListOfEnclosingTilePos()
val incompleteTiles = tiles.filterNot { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
```
So using the padded download bbox would lead to incomplete tiles. But it isn't, so actually we never have incomplete tiles (didn't see the log message that `incompleteTiles` is not empty).

----
Putting the data of the padded area (outside the non-padded bbox) is _necessary_ if the tile already exists in cache (e.g. because an overlay is enabled).
If the data wouldn't be put, the cache would return a completely empty tile (it already existed in cache and wasn't changed), while the database would return a mostly empty tile with some data near the edge.

Reply:
> So using the padded download bbox would lead to incomplete tiles. But it isn't, so actually we never have incomplete tiles (didn't see the log message that incompleteTiles is not empty).

Ah I see, okay. So this was the misunderstanding. I wasn't meaning to talk about that the `bbox` does not align with the tiles rect, but that the `items` span a larger area than would be contained in that `bbox`/tiles rect:

1. due to the download padding, when calling `replaceAllInBBox(items: Collection<T>, bbox: BoundingBox)`, some `items` are outside the specified `bbox`.
2. hence,`replaceAllInTiles(items: Collection<T>, tiles: Collection<TilePos>)` must sort each item into their tile **but drop those that do not fit into any specified in `tiles`** 
3. And thus, my [remark earlier](https://github.com/streetcomplete/StreetComplete/pull/4125#discussion_r973586207) about the commented-out code is true - it was erroneous because it did not check whether the tile the item belongs to according to `getTilePos` is actually one of the `tiles`. That would lead to items that are outside of the `bbox` to be sorted into (new) tiles

That all, assuming that sorting the items within that padding into tiles outside the bbox would be erronous. I didn't think about whether this would actually be an error (because on the other hand, the download padding is present in the db, too), which is why I asked 

> Wait, you put data into "incomplete" tiles?

Reply:
So, I do understand that:
1. tiles completely or half within the bbox are replaced completely
2. tiles half within the bbox are removed
3. items are always added, whether in the bbox or not, but only if the tile already exists

Hence, there can be tiles that only contain the data from the "download padding" (e.g. lower edge) or multiple download paddings (e.g. lower edge + right edge) because the data is always added, never replaced.

Reply:
But then, isn't the **current** code erronous? Because it only puts the item if it is within the bounds of any of the given tiles.

Reply:
>     1. due to the download padding, when calling `replaceAllInBBox(items: Collection<T>, bbox: BoundingBox)`, some `items` are outside the specified `bbox`.
>
>     2. hence,`replaceAllInTiles(items: Collection<T>, tiles: Collection<TilePos>)` must sort each item into their tile **but drop those that do not fit into any specified in `tiles`**

Why must it drop those items?
I had that done earlier, and changed it to not drop them. Because dropping those items may lead to inconsistencies between what the db and what the cache returns for a bbox that contains the padded area.
I observed such inconsistencies when the (previously empty) tile containing the padded area was cached before the download.

> That all, assuming that sorting the items within that padding into tiles outside the bbox would be erronous. I didn't think about whether this would actually be an error (because on the one hand, the download padding is present in the db, too. On the other hand, the complete tile is replaced with whatever was handed in in`replaceAllInBBox`)

The tile is completely replaced, and
* previously: what is not in the tile(s) is ignored. This leads to inconsistencies between cache and db if such an "incomplete" was already cached before download. If the tile was not cached, it's fetched from in its "incomplete" state and db and cache are consistent.
* currently: what is not in the tile(s) is put to cache, if the containing tile exists: If this is not done, the only way to ensure a cache consistent with db is by providing the padded bbox, leading to the incomplete tiles being dropped from cache. If then they are fetched from db, the are in the same "incomplete" state as in the current version.

> So, I do understand that:
> 
>     1. tiles completely within the bbox are replaced completely
> 
>     2. tiles half within the bbox are removed
> 
>     3. items are always added, whether in the bbox or not, but only if the tile already exists
>
> Hence, there can be tiles that only contain the data from the "download padding" (e.g. lower edge) or multiple download paddings (e.g. lower edge + right edge) because the data is always added, never replaced.

Yes, this is what currently happens.
And this is the data as it is in the database, so actually it **must** happen.

> But then, isn't the current code erronous? Because it only puts the item if it is within the bounds of any of the given tiles.

You're right, I didn't notice this...
I didn't repeat the situation that made me find the inconsistency with the current version.
I actually only did it with
```kotlin
 for (item in items) { 
     byTile[item.getTilePos()]?.let { 
         it.add(item) 
         byKey[item.getKey()] = item 
     } 
 } 
```
And changed to the current version because the code above made the tests fail.

I definitely meant to loop over all tiles, and was pretty sure that was actually happening until right now.

---
But... now I think looping over all tiles is **not** the correct way, because `pos in bboxByTile[tile]` does not give a unique tile:
`LatLon(0.0, 0.0) in Tile.asBoundingBox(16)` returns true for 3 tiles: TilePos(x=32768, y=32768) (which is the tile found by `getTilePos`), TilePos(x=32767, y=32768), TilePos(x=32768, y=32767)

So we need to go back to the `getTilePos` version, and make the tests work...

Reply:
> So we need to go back to the getTilePos version, and make the tests work...

OK yes, so we understand each other. Maybe you can look into why enclosingTilePos etc. seem to return the (slightest) wrong things? Otherwise, I might get to it soon, too.

Reply:
It's related to https://github.com/streetcomplete/StreetComplete/blob/890fdc4bc7e6ccadcad78e4941d056225140bb32/app/src/main/java/de/westnordost/streetcomplete/data/download/tiles/TilesRect.kt#L104
If I set it to 0.0, things seem to work for what I need here, i,e.
```kotlin
val tilesRect = listOf(LatLon(0.0, 0.0), LatLon(0.0002, 0.0002)).enclosingBoundingBox().enclosingTilesRect(16)
val tilePos = LatLon(0.0, 0.0).enclosingTilePos(16)
assertTrue(tilesRect.contains(tilePos.toTilesRect()))
```
works.

But then other tests fail, e.g. `convert bbox to tiles rect and back results in same bbox`, and basically everything that relies on that (also the cache needs it to work).

Reply:
So I switched back to `getTilePos()`, but now several tests fail.

I found that tests pass when shifting the tile edges a little (by 1e-9 degrees):
```kotlin
private fun tile2lon(x: Int, zoom: Int): Double =
    x / numTiles(zoom).toDouble() * 360.0 - 180.0 - 1e-9

private fun tile2lat(y: Int, zoom: Int): Double =
    atan(sinh(PI - 2.0 * PI * y / numTiles(zoom))).toDegrees() + 1e-9

private fun lon2tile(lon: Double, zoom: Int): Int =
    ((numTiles(zoom) * (lon + 180.0) / 360.0) + 1e-9).toInt()

private fun lat2tile(lat: Double, zoom: Int): Int =
    ((numTiles(zoom) * (1.0 - asinh(tan(lat.toRadians())) / PI) / 2.0) - 1e-9).toInt()
```
but that is not a solution. It only works because many tests use LatLon(0.0, 0.0), which is not at a tile edge with these changes.
If tests would use LatLon at the new tile edges, the same issues would arise again.

Reply:
I still haven't figured out a real solution, but the approach above might still work.
Coordinates from OSM only contain 7 decimal digits behind the comma, so anything at higher precision will never be used. 
So if we can make sure that tile edges never align exactly with coordinates from OSM, we could at least avoid the problem.

Reply:
The edges of a `BoundingBox`es are inclusive (a bounding box of 0,0 to 1,1 contains both position 0,0 and 1,1) while the edges of `TilePos`itions/`TileRect` are inclusive only on the lower two edges but not on the upper two edges (a tile pos covering the area of 0,0 to 1,1 contains 0,0 and only 0.999...,0.999... but not 1,1.

So I think it is not correct to add the `1e-9`s there but at the conversion point of `TilePos`/`TilesRect` to bounding boxes. I'll prepare a PR...